src/main/java/org/jboss/staxmapper/XMLMapperImpl.java

```diff
@@ -100,10 +108,24 @@ public void deparseDocument(XMLContentWriter contentWriter, XMLStreamWriter stre
         contentWriter.writeContent(new FormattingXMLStreamWriter(streamWriter));
     }
 
-    @SuppressWarnings({ "unchecked" })
+    private XMLElementReader<?> getParser(final QName name) {
+        rootElements.computeIfAbsent(name, qName -> {
+            XMLElementReader<?> reader = rootElementsProvider.get(name).get(); //lazy create of parser
+            return new WeakReference<>(reader);
```

Comment:
The "all versions of all subsystems" bit isn't as relevant to the WeakReference question. Simply adding your provider concept ensures that only relevant parsers are created.

If there's evidence the memory savings from discarding the actually used parsers is significant, conceivably we could just not cache parsers obtained from the providers at all and hold no reference outside of the stack. AFAICT the only place that would affect anything is on the DC when it parses domain.xml and needs the same parser in multiple profiles.